### ISSUE TITLE
Implementation of Focal Cutout or Focal Masking

### ISSUE THREAD

**Issue by sayakpaul** [keras-team/keras-cv#458]
Not sure if any other work has implemented and investigated this approach of Focal Masking before but [1] combines Focal Masking and Random Masking to improve self-supervised pre-training for learning visual features. 

The idea of Focal Masking (and comparison to Random Masking) is demonstrated in the figure below:

<img width="1236" alt="image" src="https://user-images.githubusercontent.com/22957388/170862493-fdb86cd4-e359-4ee5-b2b7-5b3a8468bcbc.png">
<sup>Source: [1]</sup><br>

Masking strategies have been super useful for NLP (BERT). They have shown good promise for Vision too (Masked Autoencoders, BeiT, SimMIM, Data2Vec, etc.). I believe it'd be useful for KerasCV users because it would allow them to pre-train models using different masking strategies and investigate their impact. 

Random Masking is quite similar to Cutout ([TensorFlow Addons module](https://www.tensorflow.org/addons/api_docs/python/tfa/image/random_cutout)) and that is why I used the term Focal Cutout. 

**References**

[1] Masked Siamese Networks: https://arxiv.org/abs/2204.07141

/cc: @LukeWood

**Comment by LukeWood** [comment]
Thanks Sayak!  Looks like a great tool.  I'd love to be able to reproduce masked autoencoders using KerasCV components.  Thanks!

**Comment by ariG23498** [comment]
Hey folks,

I took a look at the paper and the code. These are my thoughts:

- @AdityaKane2001 rightly pointed out
  > in their official implementation ([here](https://github.com/facebookresearch/msn/blob/4388dc1eadbe3042b85d3296d41b9b207656e043/src/data_manager.py#L95-L105)) they have simply used inception crop and other augmentations
- The masking actually happens [inside the `encoder`](https://github.com/facebookresearch/msn/blob/main/src/deit.py#L250-L255)
- So we first augment the images and them mask the patches accordingly. But I am still not sure why the authors have used `RandomResizedCrop` as an augmentation before masking. This part does not really sit with what it written in the paper and the diagram shown.

I would love to chime in for this augmentation layer if we ever wanted to build it in keras-cv.

**Comment by sayakpaul** [comment]
Thanks for your interest and for pointing that out, @ariG23498! 

@AdityaKane2001 is actually implementing this as a part of GSoC, so I'm not too sure about the scope of taking another contributor on this one.

**Comment by ariG23498** [comment]
I understand! Can we have a label for `GSoC` so that we point out which issues are already taken?

@AdityaKane2001 all the very best for your contributions!

**Comment by sayakpaul** [comment]
@LukeWood let's create a label for GSoC-22-related stuff to better separate them?

**Comment by AdityaKane2001** [comment]
@ariG23498 

Thanks for the insight! I just have a small question. It looks to me that patches are getting masked randomly, and not in the order mentioned in the paper. As you said, the question of RandomResizedCrop also remains, since they have used it with a very low area factor (0.05 to 0.3) which seems quite unusual.

https://github.com/facebookresearch/msn/blob/main/src/deit.py#L250-L255

**Comment by sayakpaul** [comment]
I guess it's best to clarify this with the authors of the paper. 

@imisra @MidoAssran could you please help us with the doubts mentioned in https://github.com/keras-team/keras-cv/issues/458#issuecomment-1155090357 and https://github.com/keras-team/keras-cv/issues/458#issuecomment-1155985462?

Thanks!

**Comment by MidoAssran** [comment]
Hi @AdityaKane2001, focal masking is just extracting a small-crop from an image-view. An image-view is created via random-data augmentations of an image. For efficiency, you can do both simultaneously in the data-loader with the RandomResizedCrop function in Pytorch.

*Key points:*
Notice that the crop-scale is very small (0.05, 0.3); meaning we are extracting crops that range between 5% to 30% of the total image size, and then resize these to 96x96 pixels for efficient batch processing (so that all the focal crops can be processed in the same forward pass). This is simply equivalent to called RandomResizedCrop with the aforementioned scale and crop-size!

*On random masking:*
Random masking on the other hand cannot be implemented in the same way, since it corresponds to dropping non-contiguous patches. Therefore, after creating an image-view, the random masking is executed in the encoder by randomly dropping input patches.

This code, [here](https://github.com/facebookresearch/msn/blob/main/src/deit.py#L221-L228), ensures that patch-dropping only happens to the random mask views, and not the focal views (which were already masked in the data-loader).

**Comment by AdityaKane2001** [comment]
@MidoAssran 

Thanks for the clarification! I just have one question. The illustration in the paper suggests involvement of patching, and contiguous patches in a grid are retained while dropping the rest of the image in the case of focal masking. However, the code as well as the procedure you mentioned does not take this into consideration. Could you please share your thoughts on this?

**Comment by AdityaKane2001** [comment]
@MidoAssran 

Thanks a lot for the clarification. It is clear now.

**Comment by sachinprasadhs** [comment]
Thanks for reporting the issue! We have consolidated the development of KerasCV into the new KerasHub package, which supports image, text, and multi-modal models. Please read https://github.com/keras-team/keras-hub/issues/1831. KerasHub will support all the core functionality of KerasCV.

KerasHub can be installed with !pip install -U keras-hub. Documentation and guides are available at [keras.io/keras_hub](https://keras.io/keras_hub/).

With our focus shifted to KerasHub, we are not planning any further development or releases in KerasCV. If you encounter a KerasCV feature that is missing from KerasHub, or would like to propose an addition to the library, please file [an issue with KerasHub](https://github.com/keras-team/keras-hub/issues).
